src/sagemaker_containers/_mpi.py

```diff
@@ -0,0 +1,236 @@
+
+        if procs:
+            return procs
+        time.sleep(1)
+
+
+class MasterRunner(_process.ProcessRunner):
+    """Responsible to prepare MPI distributed training and syncronize work with the Workers.
+    """
+
+    def __init__(self, user_entry_point, args, env_vars, master_hostname, hosts, process_per_host,
+                 custom_mpi_options, network_interface_name, interval=1,
+                 timeout_in_seconds=60 * 60):
+
+        super(MasterRunner, self).__init__(user_entry_point, args, env_vars)
+
+        self._master_hostname = master_hostname
+        self._hosts = hosts
+        self._process_per_host = process_per_host
+        self._custom_mpi_options = custom_mpi_options
+        self._network_interface_name = network_interface_name
+        self._interval = interval
+        self.timeout_in_seconds = timeout_in_seconds
+
+    def _setup(self):  # type: () -> None
+        logger.info('Starting MPI run as worker node.')
+        logger.info('Creating SSH daemon.')
+        _start_sshd_daemon()
+
+        self._wait_for_workers()
+
+    def _wait_for_workers(self):  # type: () -> None
+        logger.info('Waiting for MPI workers to establish their SSH connections')
+
+        workers = [host for host in self._hosts if host != self._master_hostname]
+        with _timeout.timeout(seconds=self.timeout_in_seconds):
+            for host in workers:
+                while not _can_connect(host):
+                    time.sleep(self._interval)
+                logger.info('Worker %s available for communication', host)
+
+    def _create_command(self):  # type: () -> List[str, Any]
+        num_hosts = len(self._hosts)
+        num_processes = self._process_per_host * num_hosts
+
+        # By default, use one process per GPU, or one process per node (if training with CPU).
+        if self._process_per_host == 1:
+            host_list = self._hosts
+        else:
+            host_list = ['%s:%s' % (host, self._process_per_host) for host in self._hosts]
+
+        msg = 'Env Hosts: %s Hosts: %s process_per_hosts: %s num_processes: %s'
+        logger.info(msg, self._hosts, host_list, self._process_per_host, num_processes)
+
+        overridden_known_options, additional_options = _parse_custom_mpi_options(
+            self._custom_mpi_options)
+
+        logger.info("Network interface name: %s" % self._network_interface_name)
+
+        # TODO(mvs): explain MPI setttings
+        command = ['mpirun',
+                   '--host', ','.join(host_list),
+                   '-np', str(num_processes),
+
+                   '--allow-run-as-root',
+                   '--display-map',
+                   '--tag-output',
+
+                   '-mca', 'btl_tcp_if_include', self._network_interface_name,
+                   '-mca', 'oob_tcp_if_include', self._network_interface_name,
+                   '-mca', 'plm_rsh_no_tree_spawn', '1',
```

Comment:
Are all of these necessary? I think we should bias towards not including options rather than including options that may not be necessary.

A few suspect ones are plm_rsh_no_tree_spawn, NCCL_DEBUG, and NCCL_MIN_NRINGS.